### PR TITLE
Add SQLite (with WRITE support)

### PR DESCRIPTION
**Pull request by VitaSmith**
The current SDK provides only support for the default SQLite implementation
from Sony, which is locked down to read-only and is therefore pretty limited.

This patch adds full R/W support, through a new libsqlite.a library, as well
as 2 new calls: sqlite3_rw_init() and sqlite3_rw_exit().

See https://github.com/VitaSmith/libsqlite for details and samples.

Additional Notes:
-------------------

* Note that `make install` of libsqlite is designed to __override__ the existing `include\psp2\sqlite.h` (to add the `sqlite3_rw_init()` and `sqlite3_rw_exit()` declaration, as I don't think it would make much sense to require SQLite Vita users to use 3 different headers (one for `sceSqliteConfigMallocMethods()`, one for the new `sqlite3_rw_###` and one last for the regular `sqlite3.h`). If this proposal is integrated, I can submit a patch against _vita-headers_ to make the `sqlite.h` modification permanent, in which case I'll modify `make install` to just copy the library.
* The way R/W support is achieved is by overriding the VFS, with just the minimal subset of what we need (`xOpen()`, `xDelete()` and `xWrite()`), while leaving as much as possible of the original VFS untouched.

### CODE REVIEW

**Comment by VitaSmith** [comment]
Any update on including this?

As mentioned in TheOfficialFloW/VitaShell#323, not having libsqlite included in VitaSDK is kind of preventing me from wanting to push a required update to VistaShell...

**Comment by devnoname120** [comment]
> Note that `make install` of libsqlite is designed to override the existing `include\psp2\sqlite.h`

I'm not fond of this behavior, maybe we should rename this file in the vitasdk?

**Comment by devnoname120** [comment]
@VitaSmith I'm sorry I read your message too fast, please discard the first part of my previous message.

I suggest you to rename your `sqlite.h` to something like `sqlite_rw.h` and put it in `arm-vita-eabi/include` (rename the lib too).

Users won't need to include `arm-vita-eabi/include/psp2/sqlite.h` since the methods are called by your library behind the hood, it's transparent for the user. You can add a warning to tell users to use your library instead.

So to sum up, users will need to:
- Include `arm-vita-eabi/include/sqlite_rw.h`.
- Include `arm-vita-eabi/include/sqlite.h`.
- Link `sqlite_rw`.
- Link `SceSqlite_stub`.
- Run some functions like in your sample.

**Comment by VitaSmith** [comment]
@devnoname120, thanks for the reply.

I understand your points, but I'd very much prefer to work something out that merges the declarations into a single `arm-vita-eabi/include/sqlite.h` rather than add a `_rw` version of it. There's so little content that would go into these separate headers that it really makes very little sense to me to try to split them further...
And again, I am more than willing to submit a patch against _vita-headers_ to remove the need to override the current `sqlite.h`. However, the trick is that, since these are officially separate projects, one of these changes need to happen before the other, which is why I preferred to start here.

Also, I'm not fond of at all with renaming `sqlite` to `sqlite_rw`.
Obviously, when one links with `sqlite`, one expects that the library will not be restricted, and that linking for the Vita should be very similar to linking for other platforms, that just use something like `-lsqlite` to access a R/W version of the library. If there was any possibility of conflict with an existing Vita library, I would understand trying to steer away from using `sqlite`, but the fact of the matter is that, if you want to use the native RO version, you wouldn't link with `-lsqlite`, just `-lSceSqlite_stub`. So I don't really see the point of trying to preserve the `sqlite` library name here. Or, in case this is your concern, the end goal here is that anyone who is already using their own compiled version of sqlite as a library in their app, and might therefore be using `-lsqlite` in their project, should be able to use the proposed libsqlite as a drop in replacement, and remove their requirement to compile their own version, which is precisely the point of trying to get it integrated into packages.

As such, adding a suffix to the library, as if to suggest it isn't the full featured version of SQLite somehow, is likely to create more confusion and make developers wonder whether they can actually expect to use the library as a drop-in replacement.

What I can do then, if you agree, is this:
* Submit a patch for `sqlite.h` against  _vita-headers_, to add the 2 init/exit declarations required by libsqlite.
* Once this has been integrated, alter libsqlite so that it doesn't modify headers on install, and resubmit this request.

**Comment by CelesteBlue-dev** [comment]
Sorry, so you are going to add a sqlite library that comes from where ?
@xyzz already ported sqlite for exemple in henkaku offline-installer.
Is that different ?

**Comment by VitaSmith** [comment]
> Is that different ?

Yes, because the firmware does provide an SQLite library, that is read-only by default, but that can be unlocked for full R/W. In other words, you should be able to completely eliminate the need to compile your own libsqlite, and just rely on the the one that's already there.

**Comment by VitaSmith** [comment]
Maybe I should also clarify that, if you look at `libSceSqlite_stub.a`, you'll find that the `sqlite3_####` references from the firmware are already provided there, which is why we don't need to link with anything else to be able to use these calls. So, if all you need to do is read data from an SQLite3 DB, you can just link with `SceSqlite_stub`, provide your `sqlite3.h` header, and you're good to use the internal library.

The only issue is that the underlying sqlite3 VFS (Virtual File System) that is used by the native library appears to have secured down by Sony, to prevent some write operations (mostly creation of new files). But we can work around that by updating the VFS, which is what libsqlite does.

At any rate, there is a [sample](https://github.com/VitaSmith/libsqlite/tree/master/sample) in you can have a look at, if you want a better understanding of how this is meant to work in practice.

**Comment by devnoname120** [comment]
@VitaSmith Can you come on the [vitasdk room](https://riot.im/app/#/room/#vitasdk:matrix.org) so that we can discuss about your pull request?

**Comment by VitaSmith** [comment]
Is there anything specific you'd like to discuss that can't be done on the issue tracker? I must say I have a strong preference for the issue tracker, because it keeps the discussion as visible as possible, and also it doesn't require people to be online at the exact same time (which, as you can see, is not something I can commit to being able to do, as I don't check on this community on regular basis). Otherwise, please indicate a good time (GMT) to be in the vitasdk room, as it was empty when I last tried it.

**Comment by devnoname120** [comment]
I like to use direct chat for discussions that otherwise might be endless, I feel that this is the case here.
A good time for me is any day between 18:30 and 23:59 UTC+1.

Don't hesitate to highlight me by writing `noname120[m]` on the chat so that I know you're connected. 😃

**Comment by VitaSmith** [comment]
Cancelled as per discussion.
